infrastructure/comms/src/connection/net_address/parser.rs

```diff
@@ -0,0 +1,207 @@
+//  DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+//  SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+//  SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+//  WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+//  USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+use super::{I2PAddress, OnionAddress};
+
+/// Provides simple parsing functionality for address strings.
+/// Currently, it contains parsing implementations for Onion and I2P.
+pub(crate) struct AddressParser<'a> {
+    pos: usize,
+    data: &'a [u8],
+}
+
+impl<'a> AddressParser<'a> {
+    /// Create a new address parser
+    pub fn new(s: &'a str) -> Self {
+        AddressParser { pos: 0, data: s.as_bytes() }
+    }
+
+    /// Parse I2P address
+    pub fn parse_i2p(&mut self) -> Option<I2PAddress> {
+        self.read_atomic(|p| {
+            let name = match p.read_base32_string() {
+                Some(n) => {
+                    if n.len() != 52 {
+                        return None;
+                    }
+                    n
+                },
+                None => return None,
+            };
+
+            match p.read_until_end() {
+                Some(s) => {
+                    if s.to_ascii_lowercase() != b".b32.i2p" {
+                        return None;
+                    }
+                },
+                None => return None,
+            }
+
+            Some(I2PAddress { name })
+        })
+    }
+
+    /// Parse Onion address
+    pub fn parse_onion(&mut self) -> Option<OnionAddress> {
+        self.read_atomic(|p| {
+            let public_key = match p.read_base32_string() {
+                Some(pk) => {
+                    // Valid onion address lengths
+                    if pk.len() != 16 && pk.len() != 56 {
+                        return None;
+                    }
+
+                    pk
+                },
+                _ => return None,
+            };
+
+            match p.read_until_char(':') {
+                Some(buf) => {
+                    if buf.to_ascii_lowercase() != b".onion" {
+                        return None;
+                    }
+                },
+                None => return None,
+            }
+
+            if p.consume_char(':').is_none() {
+                return None;
+            }
+
+            let port = match p.read_number() {
+                Some(p) => p,
+                None => return None,
+            };
+
+            if port > std::u16::MAX as u64 {
+                return None;
+            }
+
+            Some(OnionAddress { public_key, port: port as u16 })
+        })
+    }
+
+    fn is_base32_char(&self, ch: char) -> bool {
+        if ch >= 'a' && ch <= 'z' {
```

Comment:
Doesn't base32 use uppercase "A" to "Z"?

Reply:
You are right, I think it's better to use upper case